### PR TITLE
Add a tutorial for running MNIST with Node.js

### PR DESCRIPTION
**Pull request by nkreeger**
This is a first pass at introducing a small demo for running MNIST on Node.js. It doesn't get into details about the MNIST model (we already have a demo for that) - it mostly highlights how to load the Node.js bindings, set the backend, and load data using Node.js APIs instead of browser-specific ones.

Please take a crack at my copy and let me know if this structure doesn't flow.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/161)

### CODE REVIEW

**Comment by caisq** [comment]
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[source/tutorials/mnist-node.md, line 2 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqPezfAvhMAtJ1ugWY:-LDqPezfAvhMAtJ1ugWZ:b-yxf877) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L2)):*
> ```
> mnist
> ```

Maybe be more specific: like mnist_nodejs?

---

*[source/tutorials/mnist-node.md, line 8 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqSDYUDEqlD8zS6Mjh:-LDqSDYUDEqlD8zS6Mji:b-lhz6yt) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L8)):*
> ```
>  we'll show how to utilize the power of TenorFlow's C library with TensorFlow.j
> ```

I would mention the word "node.js" in the opening sentence.

---

*[source/tutorials/mnist-node.md, line 27 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqSPdq9tsFoBEeJqvM:-LDqSPdq9tsFoBEeJqvN:btz1jb1) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L27)):*
> ```
> npm install
> ```

Maybe mention somewhere below that `yarn` works as well?

---

*[source/tutorials/mnist-node.md, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqSX4q9P72EhPM_c7G:-LDqSX4q9P72EhPM_c7H:b-t3m4xr) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L35)):*
> ```
> TensorFlow's C 
> ```

You might want to be a little specific: this is the eager C API library (as versus the graph one).

---

*[source/tutorials/mnist-node.md, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqSjct549y3Va8Vpfu:-LDqSjct549y3Va8Vpfv:bmk4ab4) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L35)):*
> ```
> high performance.
> ```

Regarding high performance, you can mention the multi-threaded (Eigen) backend for CPU and CUDA-accelerated backend for GPU. 

---

*[source/tutorials/mnist-node.md, line 67 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqT8VY1Vv0muuMkpmd:-LDqT8VY1Vv0muuMkpme:b3f5htq) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L67)):*
> ```
> embedded
> ```

Not clear to me what the word "embedded" means here. Maybe remove it.

---

*[source/tutorials/mnist-node.md, line 71 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqTEvS7HpwtRXhf41z:-LDqTEvS7HpwtRXhf42-:bbv9m1b) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L71)):*
> ```
> today
> ```

The word "today" seems to be a little out of context. Nowhere in this tutorial is the past or the future talked about.

---

*[source/tutorials/mnist-node.md, line 98 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqTgF23I-kGZD8NhHh:-LDqTgF23I-kGZD8NhHi:b-xz2y4) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L98)):*
> ```
> image samples
> ```

--> pixels

---

*[source/tutorials/mnist-node.md, line 100 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqTYnFCP3omcK1b3lM:-LDqTYnFCP3omcK1b3lN:bgyatau) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L100)):*
> ```
> 1.0
> ```

Are you sure the lower bound is not 0.0?

---

*[source/tutorials/mnist-node.md, line 132 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqV28b67KoRQDvoJoW:-LDqV28b67KoRQDvoJoX:b-c0g4hz) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L132)):*
> ```
> }
> ```
> ```

Is this tutorial incomplete as it stands?  So far it talks about only how to load and preprocess the data. It doesn't talk about defining, training or evaluating the model.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/161)*

**Comment by tafsiri** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[source/tutorials/mnist-node.md, line 8 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqWPtY1hQFHZ6MyQtA:-LDqWPtY1hQFHZ6MyQtB:b-9z0p30) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L8)):*
> ```Markdown
> # Running TensorFlow.js on Node
> 
> In this tutorial, we'll show how to utilize the power of TenorFlow's C library with TensorFlow.js. We'll be using the MNIST convolutional neural
> ```

Typo in first mention of TensorFlow.
I'd also say "We'll be using a convolutional neural network in this tutorial ..." so as not to imply that there is a singular MNIST convolutional neural net.

---

*[source/tutorials/mnist-node.md, line 9 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqWWl0-ZaZgTFy9j5t:-LDqWWl0-ZaZgTFy9j5u:b-ozg05c) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L9)):*
> ```Markdown
> 
> In this tutorial, we'll show how to utilize the power of TenorFlow's C library with TensorFlow.js. We'll be using the MNIST convolutional neural
> network in this tutorial - but loading data from binary files instead of the web-based `fromPixels()` method.
> ```

I would not mention the 'fromPixels' method here as it hasn't been described yet. Instead I think you can say something like 'this tutorial will focus primarily focus on the code that is node specific , including installation of the node bindings and loading data from disk'.

---

*[source/tutorials/mnist-node.md, line 13 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqX0ZI6bIFlBRn2gDC:-LDqX0ZI6bIFlBRn2gDD:bboefat) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L13)):*
> ```Markdown
> ## Prerequisites
> 
> This tutorial assumes prior knowledge of fundamental concepts of TensorFlow.js, MNIST dataset, and Node.js. You'll also need a machine that supports core TensorFlow. See [these directions](https://www.tensorflow.org/install/) for machine and hardware requirements.
> ```

core Tensorflow -> TensorFlow.

---

*[source/tutorials/mnist-node.md, line 33 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqY5zBC6oxQ5iJmM7k:-LDqY5zBC6oxQ5iJmM7l:by20yhy) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L33)):*
> ```Markdown
> This application downloads, trains, and evaluates against the MNIST dataset. Every 20 steps, a console log shows the current loss and accuracy of the model. At the end off each pass through the training data (called an 'epoch'), the test data is run for evaluation to see how accurate the model is.
> 
> ## Loading the TensorFlow Node.js bindings:
> ```

I'd make the header title match the one below. Either have them both be "Loading the ..." or "Installing ...".  I'd also include the npm install command for the cpu bindings just like you do for the GPU one.

---

*[source/tutorials/mnist-node.md, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqYK_QCIHWrManL0pf:-LDqYK_QCIHWrManL0pg:bhhu360) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L35)):*
> ```
> this tutorial is powered by TensorFlow's C library for high performance.
> ```

I'd try and explain what this means. Instead of powered by, i'd describe that once the backend is set, tfjs ops will be executing using the C implementation of TensorFlow (also is it C or C++? Just so that we can be consistent).

---

*[source/tutorials/mnist-node.md, line 56 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqZAC_Dmy6mqJ1ZsmX:-LDqZAC_Dmy6mqJ1ZsmY:b-b954g6) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L56)):*
> ```Markdown
> ```js
> const tf = require('@tensorflow/tfjs');
> // require('@tensorflow/tfjs-node');
> ```

I'd add a text comment explaining that we do not need to load tfjs-node. Basically explain why this line is commented out. You could also just remove this line.

---

*[source/tutorials/mnist-node.md, line 67 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqZhYw2Aomzx2zfth4:-LDqZhYw2Aomzx2zfth5:bn4lkbc) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L67)):*
> ```Markdown
> The [browser tutorial for MNIST](mnist.html) uses one large image with all the training data embedded inside. The `tf.fromPixels()` method is used to convert HTML image data to Tensors. In Node.js, we will download binary files that contain training and test information for our model.
> 
> Each MNIST dataset is stored in two different files - one containing an embedded representation of each pixel in the image and another file that contains the label for each image. In this tutorial, we will download and use a training and test set for a combination of 4 files.
> ```

Its unclear what the multiple MNIST datasets are in this sentence.

---

*[source/tutorials/mnist-node.md, line 69 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDq_H4PDqMnZekbLpBR:-LDq_H4PDqMnZekbLpBS:b-dvfdt8) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L69)):*
> ```Markdown
> Each MNIST dataset is stored in two different files - one containing an embedded representation of each pixel in the image and another file that contains the label for each image. In this tutorial, we will download and use a training and test set for a combination of 4 files.
> 
> ## Parsing Training and Test Data
> ```

Should this be a third level heading?

---

*[source/tutorials/mnist-node.md, line 74 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDq_Pn841WI9Cp6Xk4-:-LDq_Pn841WI9Cp6Xk40:b-lmi53i) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L74)):*
> ```Markdown
> 
> ```js
> async function fetchOnceAndSaveToDiskWithBuffer(filename) {
> ```

does this code parse the data? It seems to just unzip and save it.

---

*[source/tutorials/mnist-node.md, line 96 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDq_Wog2EU_PDnXz7av:-LDq_Wog2EU_PDnXz7aw:b-tzlljb) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L96)):*
> ```Markdown
> ```
> 
> ## Converting binary data to Tensors
> ```

I suggest a third level header

---

*[source/tutorials/mnist-node.md, line 98 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDq_nzZCRdQqgyt17lg:-LDq_nzZCRdQqgyt17lh:bw1e5a4) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L98)):*
> ```
> labels file
> ```

It might be nice to list out what the downloaded files are (possibly in the previous section)? At this point if i haven't actually run the code its hard to keep up with the context this is written from.

---

*[source/tutorials/mnist-node.md, line 100 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqarPiAJerb5F9sWu8:-LDqarPiAJerb5F9sWu9:b-vhuuey) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L100)):*
> ```
> `768`
> ```

Is this bytes? I notices it is different from 784 bytes mentioned earlier. Overall this sentence is somewhat unclear, is the Buffer responsible for converting the chunks to Float32Array or is the API.  I think it might be easier to show the annotated code first then explain it a bit, then show the label loading code then explain the difference. 

---

*[source/tutorials/mnist-node.md, line 132 at r1](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-LDqV28b67KoRQDvoJoW:-LDqcYjq2eEEr38ljxUv:b15wq6d) ([raw file](https://github.com/tensorflow/tfjs-website/blob/3905168b111781b4504f8477cbc398815f67996a/source/tutorials/mnist-node.md#L132)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

> ```
> }
> ```
> ```

Is this tutorial incomplete as it stands?  So far it talks about only how to load and preprocess the data. It doesn't talk about defining, training or evaluating the model.
</blockquote></details>

I think at the least it should reiterate that the rest of the code is similar to the browser based one and possibly show all the code inline if its shortish.  

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/161)*

**Comment by tafsiri** [comment]
I think overall id suggest adding one extra level of hiearchy in the the titles something like

# Main title

## Installation & Loading

### CPU Bindings

### GPU Bindings

## Loading MNIST data from disk

### Getting the data

### Parsing the data

---

Review status: 0 of 4 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-website/161#-:-LDqe5hL92C5sYSEh5ux:b-yzbyou)*
